### PR TITLE
Webhook Publishing Update

### PR DESCRIPTION
**Pull request by bsnuggs1**
UPDATED publish to make fields required
ADDED webhook publishing
ADDED webhook validation

### CODE REVIEW

File: actions/jenkinsfile/jenkinsfile.go [hunk 1]
```diff
@@ -100,23 +242,99 @@ func Publish(r *http.Request) api.Response {
...
-			}
-			return api.Response{Error: errors.Err("failed to publish Jenkinsfile"), Data: e}
-		}
 	}
+
 	return api.Response{Data: "OK"}
 }
+
+// Checks to see if we have the DQCI webhook created on the repository.  If not, we create one.
+func PublishWebhooks(r *http.Request) api.Response {
```
Comment (tiger5226) [inline actions/jenkinsfile/jenkinsfile.go]: why is this exported from the package?

File: actions/routes.go [hunk 2]
```diff
@@ -28,6 +28,7 @@ func GetRoutes() *Routes {
 	routes.Set("/bucket/list", List)
 	routes.Set("/jenkinsfile/list", jenkinsfile.List)
 	routes.Set("/jenkinsfile/publish", jenkinsfile.Publish)
+	routes.Set("/jenkinsfile/publish/webhooks", jenkinsfile.PublishWebhooks)
```
Comment (tiger5226) [inline actions/routes.go]: there should only be two. `/<resource>/<action`. I would call it `/jenkins_hooks/publish` and change the other function to `PublishJenkinsfile`.

File: actions/jenkinsfile/jenkinsfile.go [hunk 3]
```diff
@@ -51,21 +77,137 @@ func List(r *http.Request) api.Response {
 	return api.Response{Data: files}
 }
 
+func GenerateWebhook(projectKey string, repo string, hookURL string) error {
```
Comment (tiger5226) [inline actions/jenkinsfile/jenkinsfile.go]: why is this function exported from the package?

File: actions/jenkinsfile/jenkinsfile.go [hunk 4]
```diff
@@ -51,21 +77,137 @@ func List(r *http.Request) api.Response {
...
+		_, err := HandleResponseBasic(resp, "failed to create webhook")
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func ListWebhooks(projectKey string, repo string) ([]Webhook, error) {
```
Comment (tiger5226) [inline actions/jenkinsfile/jenkinsfile.go]: same thing, why is this exported? It appears to be a helper method.

File: actions/jenkinsfile/jenkinsfile.go [hunk 5]
```diff
@@ -51,21 +77,137 @@ func List(r *http.Request) api.Response {
...
+		fmt.Println("===== Webhooks =====")
+		for i, hook := range hooks {
+			fmt.Println(i, hook)
+		}
+	}
+
+	return hooks, nil
+}
+
+func ValidateBitbucketEnvs(bitbucketURL string, username string, password string) error {
```
Comment (tiger5226) [inline actions/jenkinsfile/jenkinsfile.go]: should not be exported

File: actions/jenkinsfile/jenkinsfile.go [hunk 6]
```diff
@@ -51,21 +77,137 @@ func List(r *http.Request) api.Response {
...
+	}
+
+	if len(username) == 0 || len(password) == 0 {
+		return errors.Base("unable to find credentials for bitbucket")
+	}
+
+	return nil
+}
+
 func Publish(r *http.Request) api.Response {
```
Comment (tiger5226) [inline actions/jenkinsfile/jenkinsfile.go]: Any api call, should start with the parameter validation. You should not pass the request around to other functions. The information should be extracted from the request at this level and that information passed around.

File: actions/jenkinsfile/jenkinsfile.go [hunk 7]
```diff
@@ -100,23 +242,99 @@ func Publish(r *http.Request) api.Response {
...
+	// we didn't exit yet, so we'll need to generate the webhook
+	err = GenerateWebhook(projectKey, repo, hookURL)
+	if err != nil {
+		return api.Response{Error: errors.Err(err)}
+	}
+
+	return api.Response{Data: "OK"}
+}
+
+// Checks to see if we have a matching url in a set of webhooks
```
Comment (tiger5226) [inline actions/jenkinsfile/jenkinsfile.go]: Go comments should start with a space then the name of the function so:
```
// Contains checks to see if...
```

File: actions/jenkinsfile/jenkinsfile.go [hunk 8]
```diff
@@ -100,23 +242,99 @@ func Publish(r *http.Request) api.Response {
...
+
+	if len(value) > 0 {
+		err = nil
+	}
+
+	return value, err
+}
+
+// Simple response handler, prints some simple infomration and will fail if any status is 300 or greater
+func HandleResponseBasic(resp *http.Response, failMessage string) ([]byte, error) {
```
Comment (tiger5226) [inline actions/jenkinsfile/jenkinsfile.go]: I just through this together, probably a good library for it.

File: actions/routes.go (hunk 2 above)
Comment (bsnuggs1) [reply actions/routes.go]: Hey Beamer, I had avoided changing the jenkinsfile/publish route so I could avoid having to update the UI for now as a quick change.

If I am going to update the routes, would it better to have 3 different ones?

- /all/publish
- /jenkinsfile/publish
- /jenkins_hook/publish 

This way we don't have to call two separate API calls if we want to just publish everything at once?  The parameters needed for both overlap.

File: actions/jenkinsfile/jenkinsfile.go (hunk 3 above)
Comment (bsnuggs1) [reply actions/jenkinsfile/jenkinsfile.go]: Doesn't need to be.  I only understood how the export/import after submitting the code (had looked at the travis ci results).

File: actions/jenkinsfile/jenkinsfile.go (hunk 4 above)
Comment (bsnuggs1) [reply actions/jenkinsfile/jenkinsfile.go]: Doesn't need to be.  I only understood how the export/import after submitting the code (had looked at the travis ci results).

File: actions/jenkinsfile/jenkinsfile.go (hunk 5 above)
Comment (bsnuggs1) [reply actions/jenkinsfile/jenkinsfile.go]: Doesn't need to be.  I only understood how the export/import after submitting the code (had looked at the travis ci results).

File: actions/routes.go (hunk 2 above)
Comment (tiger5226) [reply actions/routes.go]: Well, maybe we should just add the second part to the existing api then no changes are needed to the UI at all. Then we dont have to worry about this at all. Lets chat tomorrow if I am missing something.

Comment (bsnuggs1) [reply actions/routes.go]: I think its better that way as well.  We can make updates to separate it later.

Update is available now

**Pull request merged**